### PR TITLE
Unix: make Backspace work when Console.ReadLine has wrapped lines

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/corefx/issues/37056

There are a number of things that should work:

```
Console.Write("Enter your name: ");
Console.ReadLine();
```
In this case we don't want backspace to delete the prompt.

And, when the user is typing at the bottom of the terminal, we don't want weird behaviour because the cursor position where editing started will move up as lines are wrapping.

I was looking for a way to avoid calling `TryGetCursorPosition` but I think it is needed to make these scenarios work. Thanks to the caching in `PalConsole`, we won't be fetching multiple times.

@stephentoub @wtgodbe ptal

### CODE REVIEW

File: src/System.Console/src/System/IO/StdInReader.cs [hunk 1]
```diff
@@ -141,7 +141,21 @@ private string ReadLine(bool consumeKeys)
...
+
+                                // The ReadLine input may wrap accross terminal rows and we need to handle that.
+                                // note: ConsolePal will cache the cursor position to avoid making many slow cursor position fetch operations.
+                                if (ConsolePal.TryGetCursorPosition(out int left, out int top, reinitializeForRead: true) &&
+                                    left == 0 && top > 0)
+                                {
+                                    // Move to end of previous line
+                                    ConsolePal.SetCursorPosition(ConsolePal.WindowWidth - 1, top - 1);
+                                    // Clear from cursor to end of the line
+                                    ConsolePal.WriteStdoutAnsiString("\x1B[K", mayChangeCursorPosition: false);
```
Comment (tmds) [inline src/System.Console/src/System/IO/StdInReader.cs]: I want to put this string in the `TerminalFormatStrings.Instance` but I'm not sure if I should put it in verbatim or via a `WellKnownStrings` enum, and how to find out the numeric value for the enum.

Comment (stephentoub) [reply src/System.Console/src/System/IO/StdInReader.cs]: This is the terminfo `el`/`cb` entry.  I believe it's numerical entry value for WellKnownStrings should be 6.

File: src/System.Console/src/System/IO/StdInReader.cs [hunk 2]
```diff
@@ -141,7 +141,21 @@ private string ReadLine(bool consumeKeys)
...
                                 }
-                                Console.Write(s_moveLeftString);
+
+                                // The ReadLine input may wrap accross terminal rows and we need to handle that.
+                                // note: ConsolePal will cache the cursor position to avoid making many slow cursor position fetch operations.
+                                if (ConsolePal.TryGetCursorPosition(out int left, out int top, reinitializeForRead: true) &&
+                                    left == 0 && top > 0)
+                                {
+                                    // Move to end of previous line
+                                    ConsolePal.SetCursorPosition(ConsolePal.WindowWidth - 1, top - 1);
```
Comment (stephentoub) [inline src/System.Console/src/System/IO/StdInReader.cs]: What happens if the window is resized?

Comment (tmds) [reply src/System.Console/src/System/IO/StdInReader.cs]: It behaves as expected when I try this using gnome terminal and terminator. @laceous [reported](https://github.com/dotnet/corefx/issues/37056#issuecomment-488530388) it is not behaving ok with Konsole. I'm not sure if we should try to fix that. If we do, only the active ReadLine will be properly wrapped according to the new window size, anything written before that will still be responsibility of the terminal emulator.

File: src/System.Console/src/System/ConsolePal.Unix.cs [hunk 3]
```diff
@@ -442,23 +442,23 @@ public static int CursorTop
         private static bool s_firstCursorPositionRequest = true;
 
         /// <summary>Gets the current cursor position.  This involves both writing to stdout and reading stdin.</summary>
```
Comment (wtgodbe) [inline src/System.Console/src/System/ConsolePal.Unix.cs]: Could you add a comment here describing the `reinitializeForRead` option?

**Comment by wtgodbe** [comment]
You may want to add a test case to https://github.com/dotnet/corefx/blob/master/src/System.Console/tests/ManualTests/ManualTests.cs for this functionality, then try running the manual tests locally on a Unix box. Assuming that works, this LGTM. Thanks!

**Comment by tmds** [comment]
@wtgodbe I've addressed your feedback.

**Comment by wtgodbe** [comment]
Looks great, thanks! Merging this now.

**Pull request merged**